### ISSUE TITLE
how can rdp_check use tls1.0 and AES ?

### ISSUE THREAD

**Issue by w0shlzheNdecaI** [SecureAuthCorp/impacket#1331]
### Configuration  
impacket version:  0.9.24
Python version:  3.7.4
Target OS:  windows 2008

Windows2008 does not support tls1.1, so I use tls1.0 to connect. However, the server OpenSSL cancels the RC4 password and can only use AES. If AES is used instead, the following problems will occur:
```
part of Traceback :
    buff = tls.recv(4096)
File "Python\Python37\lib\site-packages\OpenSSL\SSL.py", line 1865, in recv
    self._raise_ssl_error(self._ssl, result)
  File "Python\Python37\lib\site-packages\OpenSSL\SSL.py", line 1700, in _raise_ssl_error
    _raise_current_error()
  File "Python\Python37\lib\site-packages\OpenSSL\_util.py", line 55, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert internal error')]
```

part of code:
```
ctx = SSL.Context(SSL.TLSv1_METHOD)
    ctx.set_cipher_list(b'AES256')
    tls = SSL.Connection(ctx, s)
    tls.set_connect_state()
    tls.do_handshake()
auth = ntlm.getNTLMSSPType1('', '')  # , True, use_ntlmv2=True)

    ts_request = TSRequest()
    ts_request['NegoData'] = auth.getData()

    tls.send(ts_request.getData())
    buff = tls.recv(4096)
    ts_request.fromString(buff)
```

**Comment by 0xdeaddood** [comment]
Hi @w0shlzheNdecaI!

Could you try what is implemented here: #1364?

**Comment by w0shlzheNdecaI** [comment]
> Hi @w0shlzheNdecaI!
> 
> Could you try what is implemented here: #1364?

Thanks! But it still not working, phenomenon by going the same

**Comment by 0xdeaddood** [comment]
Hey @w0shlzheNdecaI! 

I figured it out. We face two problems with TLS1.0:

- First, the client renegotiation. Depending on the OpenSSL version, legacy renegotiation is [disabled by default](https://www.openssl.org/docs/manmaster/man7/migration_guide.html#:~:text=Client%2Dinitiated%20renegotiation,parameter%20as%20appropriate). So, we have to enable it using the `SSL_OP_ALLOW_CLIENT_RENEGOTIATION` flag. pyOpenSSL doesn't have this flag defined. We need to [define it](https://wiki.openssl.org/index.php/List_of_SSL_OP_Flags) and set it to the SSL Context with the function set_options():
```
SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 0x00040000
ctx.set_options(SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION)
```
- Then, we have the BEAST attack protection. This protection is done by inserting empty data fragments. That is reason why we see [packet splitting](https://github.com/SecureAuthCorp/impacket/issues/875) in TLS 1.0 connections. To address it, we need to set the `SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS` option to disable the protection:
```
ctx.set_options(SSL.OP_DONT_INSERT_EMPTY_FRAGMENTS)
```
So, in conclusion, something like this should work:
```
ctx = SSL.Context(SSL.TLS_METHOD)
ctx.set_cipher_list('ALL:@SECLEVEL=0'.encode('utf-8'))
SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 0x00040000
ctx.set_options(SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION)
ctx.set_options(SSL.OP_DONT_INSERT_EMPTY_FRAGMENTS)
tls = SSL.Connection(ctx,s)
tls.set_connect_state()
tls.do_handshake()
```

Please let me know if this works for you. I'm going to test it in other scenarios and open a PR.

**Comment by w0shlzheNdecaI** [comment]
> 嘿@w0shlzheNdecaI！
> 
> 我想到了。TLS1.0 面临两个问题：
> 
> * 首先，客户重新谈判。根据 OpenSSL 版本，[默认情况下禁用](https://www.openssl.org/docs/manmaster/man7/migration_guide.html#:~:text=Client%2Dinitiated%20renegotiation,parameter%20as%20appropriate)旧版重新协商。因此，我们必须使用`SSL_OP_ALLOW_CLIENT_RENEGOTIATION`标志启用它。pyOpenSSL 没有定义这个标志。我们需要[定义它](https://wiki.openssl.org/index.php/List_of_SSL_OP_Flags)并使用函数 set_options() 将其设置为 SSL 上下文：
> 
> ```
> SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 0x00040000
> ctx.set_options(SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION)
> ```
> 
> * 然后，我们有 BEAST 攻击保护。这种保护是通过插入空数据片段来完成的。这就是我们在 TLS 1.0 连接中看到[数据包拆分的原因。](https://github.com/SecureAuthCorp/impacket/issues/875)为了解决这个问题，我们需要设置`SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS`禁用保护的选项：
> 
> ```
> ctx.set_options(SSL.OP_DONT_INSERT_EMPTY_FRAGMENTS)
> ```
> 
> 所以，总而言之，这样的事情应该有效：
> 
> ```
> ctx = SSL.Context(SSL.TLS_METHOD)
> ctx.set_cipher_list('ALL:@SECLEVEL=0'.encode('utf-8'))
> SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION = 0x00040000
> ctx.set_options(SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION)
> ctx.set_options(SSL.OP_DONT_INSERT_EMPTY_FRAGMENTS)
> tls = SSL.Connection(ctx,s)
> tls.set_connect_state()
> tls.do_handshake()
> ```
> 
> 请让我知道这是否适合您。我打算在其他场景下测试一下，开个PR。

oh, Thanks. It effective
